flytekit/core/type_engine.py

```diff
@@ -595,6 +617,99 @@ def guess_python_type(self, literal_type: LiteralType) -> Type[list]:
             return typing.List[ct]
         raise ValueError(f"List transformer cannot reverse {literal_type}")
 
+def _add_tag_to_type(x: LiteralType, tag: str) -> LiteralType:
+    x._structure = TypeStructure(tag=tag)
+    return x
+
+class UnionTransformer(TypeTransformer[T]):
+    """
+    Transformer that handles a univariate typing.Union[T]
+    """
+
+    def __init__(self):
+        super().__init__("Typed Union", typing.Union)
+
+    def get_literal_type(self, t: Type[T]) -> Optional[LiteralType]:
+
+        try:
+            trans = [(TypeEngine.get_transformer(x), x) for x in typing.get_args(t)]
+            variants = [_add_tag_to_type(t.get_literal_type(x), t.name) for (t, x) in trans]
+            return _type_models.LiteralType(union_type=UnionType(variants))
+        except Exception as e:
+            raise ValueError(f"Type of Generic Union type is not supported, {e}")
+
+    def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], expected: LiteralType) -> Literal:
+        found_res = False
+        res = None
+        res_type = None
+        for t in typing.get_args(python_type):
+            try:
+                trans = TypeEngine.get_transformer(t)
+
+                res = trans.to_literal(ctx, python_val, t, expected)
+                res_type = _add_tag_to_type(trans.get_literal_type(t), trans.name)
+                if found_res:
+                    # Should really never happen, sanity check
+                    raise TypeError(f"Ambiguous choice of variant for union type")
+                found_res = True
+            except TypeTransformerFailedError as e:
+                logger.debug(f"Failed to convert from {python_val} to {t}", e)
+                continue
+
+        if found_res:
+            return Literal(scalar=Scalar(union=Union(value=res, stored_type=res_type)))
+
+        raise TypeTransformerFailedError(f"Cannot convert from {python_val} to {python_type}")
+
+    def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type: Type[T]) -> Optional[typing.Any]:
+        union_tag = None
+        if lv.scalar is not None and lv.scalar.union  is not None:
+            union_type = lv.scalar.union.stored_type
+            if union_type.structure is not None:
+                union_tag = union_type.structure.tag
+
+        found_res = False
+        res = None
+        res_tag = None
+        for v in typing.get_args(expected_python_type):
+            try:
+                trans = TypeEngine.get_transformer(v)
+                if union_tag is not None:
+                    if trans.name != union_tag:
+                        continue
+
+                    assert lv.scalar is not None # type checker
+                    assert lv.scalar.union is not None # type checker
```

Comment:
Assertions can be [turned off](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement).

Reply:
These assertions are already provably true since `union_tag` is not None iff `lv.scalar` and `lv.scalar.union` are not None

The type checker cannot make this inference so these are here to silence the warnings